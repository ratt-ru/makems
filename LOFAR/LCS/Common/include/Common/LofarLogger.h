//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.h 11806 2008-09-28 21:07:32Z loose $

#ifndef LOFAR_COMMON_LOFARLOGGER_H
#define LOFAR_COMMON_LOFARLOGGER_H

// \file
// Macro interface to the lofar logging package.

//# Include the correct set of macros, depending on the availability of the
//# log4cplus or log4cxx packages.
#if defined(HAVE_LOG4CPLUS)
# include <Common/LofarLog4Cplus.h>
#elif defined(HAVE_LOG4CXX)
# include <Common/LofarLog4Cxx.h>
#else
# include <Common/LofarLogCout.h>
#endif

#include <Common/Exceptions.h>

namespace LOFAR 
{

// \ingroup Common
// \addtogroup LofarLogger
  // @{


//# --------------------- Common definitions ---------------
//
// \name Common definitions 
// @{

// The package name of the current package. This variable should be defined in
// the file lofar_config.h which is generated by the lofarconf script. Here,
// we will make sure that LOFARLOGGER_PACKAGE is always defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

// The full package name is defined as the concatenation of the package name
// and an optional sub-package name that may be defined by the user.
#ifdef LOFARLOGGER_SUBPACKAGE
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE "." LOFARLOGGER_SUBPACKAGE
#else
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE
#endif

// @}

//#------------------------ Assert and FailWhen -------------------------------
//#
//# NB: THROW is defined in package-dependant file.
//
// \name Assert and FailWhen
// \note The \c DBG version of \c ASSERT and \c FAILWHEN will only be in your
// compiled code when the (pre)compiler flag \c ENABLE_DBGASSERT is defined.
// @{

// If the condition of assert is NOT met a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERT(cond) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, "Assertion: " #cond); \
	} while(0)


// If the condition of assert is NOT met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERTSTR(cond,stream) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, \
			   "Assertion: " #cond "; " << stream); \
	} while(0)

// If the condition of failwhen is met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define FAILWHEN(cond) do { \
	if (cond) THROW(::LOFAR::AssertError, "Failtest: " #cond); \
	} while(0)

#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

// @}
    
} // namespace LOFAR

#endif // file read before
