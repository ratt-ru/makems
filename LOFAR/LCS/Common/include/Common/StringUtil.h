//#  StringUtil.h: useful string manipulation methods.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StringUtil.h 13875 2009-08-31 07:47:33Z diepen $

#ifndef LOFAR_COMMON_STRINGUTIL_H
#define LOFAR_COMMON_STRINGUTIL_H

// \file
// Useful string manipulation methods

//# Includes
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>
#include <Common/lofar_vector.h>
#include <Common/Exception.h>

namespace LOFAR
{

  // Useful string manipulation methods and classes.
  namespace StringUtil
  {
    // Split a string into substrings using \c c as the separator character.
    // The result may contain empty strings, e.g. when \c str contains two or
    // more consecutive occurrences of the separations character.
    //
    // For example:
    // \code
    //    res = StringUtil::split( "aa,bb,,dd,", ',' )
    // \endcode
    //
    // would yield the following array:
    // \verbatim
    //   res[0] = "aa"
    //   res[1] = "bb"
    //   res[2] = ""
    //   res[3] = "dd"
    //   res[4] = ""
    // \endverbatim
    vector<string> split(const string& str, char c);


    // Functor to compare two strings. Strings can be compared case sensitive
    // (\c NORMAL) and case insensitive (\c NOCASE).
    // \attention This class does not handle locales properly. It does string
    // comparison the way \c strcmp and \c strcasecmp (or \c stricmp for that
    // matter) do it.
    class Compare
    { 
    public:
      // String comparison mode.
      enum Mode {NORMAL, NOCASE};

      // Constructor. Initialize the comparison criterion. Default is "normal"
      // case sensitive comparison.
      Compare(Mode mode=NORMAL) : itsMode(mode) {}

      // The comparison operator
      bool operator()(const string& s1, const string& s2) const 
      {
	if (itsMode == NORMAL) return s1 < s2;
	else return lexicographical_compare(s1.begin(), s1.end(),
					    s2.begin(), s2.end(),
					    nocaseCompare);
      }

    private:
      // Helper function to do case insensitive comparison of two chars.
      static bool nocaseCompare(char c1, char c2)
      {
	return toupper(c1) < toupper(c2);
      }

      // The current comparison mode.
      Mode itsMode;
    };

  } // namespace StringUtil

//
// formatString(format, ...) --> string
//
// The function formatString accepts printf-like arguments and returns a
// formatted string. It can be used e.g. in cout constructions:
// cout << formatString("Opening connection with host %%s", hostName);
//# In real life this must be %s ofcourse but doxygen need a double %%.
const string formatString(const	char* format, ...);

//
// timeString(aTime [, gmt, format]) --> string
//
// The function timeString format the given timestamp into a human-readable
// format. The default format is yyyy-mm-dd hh:mm:ss
const string timeString(time_t     aTime, 
						bool       gmt = true,
						char*      format = "%F %T");

// Skip leading whitespace (blanks and horizontal tabs) starting at st.
// It returns the position of the first non-whitespace character.
// It returns end if all whitespace.
// It can be used in combination with rskipws.
uint lskipws (const string& value, uint st, uint end);

// Skip trailing whitespace (blanks and horizontal tabs) starting at end.
// It returns the position after the last non-whitespace character, thus
// value.substr(st, end-st) extracts the significant value.
// It returns st if all whitespace.
uint rskipws (const string& value, uint st, uint end);

// Skip past a quoted string.
// The quote character is the first character (at position st).
// Usually the quote character is ' or ", but it could be any other character.
// An exception is thrown if no ending quote character is found.
uint skipQuoted (const string& str, uint st);

// Skip past the end of a balanced pair of delimiters where nested pairs
// are also skipped. Delimiters in quoted parts are ignored.
// The starting delimiter is the first character in the string (at position st).
// The ending delimiter is given as an argument.
// The function works well if starting and ending delimiter are the same.
// An exception is thrown if the delimiters are not balanced.
// For example, it will skip something like '[[1,2,3],[4,5,6]]'.
uint skipBalanced (const string& str, uint st, uint end, char endChar);

//
// rtrim(char* CString [,len])
//
// Skip trailing spaces. If len of string is already known at invocation
// it may be given thus avoiding a relative expensive strlen call.
//
// Returns the length of the new string
//
// NOTE: original string is truncated!
//
int32 	rtrim(char*	aCstring, int32		len = 0, char* whiteSpace = " 	");

//
// char* ltrim(char*	Cstring)
//
// Skip leading spaces. A pointer to the first non-whitespace char is
// returned. (points into original string)
char*	ltrim(char*	aCstring, char* whiteSpace = " 	");

//
// rtrim(aString)
//
// Removes trailing whitespace from the string.
//
inline void rtrim(string&		aString, const string& whiteSpace = " 	")
{
	aString = aString.erase(aString.find_last_not_of(whiteSpace)+1);
}

//
// ltrim(aString)
//
// Removes leading whitespace from the string.
//
inline void ltrim(string&		aString, const string&	whiteSpace = " 	")
{
	aString = aString.erase(0, aString.find_first_not_of(whiteSpace));
}

// Return an uppercased string of \a str.
std::string toUpper(string str);

// Return a lowercased string of \a str.
std::string toLower(string str);

// @name Convert numeric value to string
// Convert the value of any of the fundamental arithmetic data types to a
// string representation. Most of the toString() methods provide the user with
// a means to override the default formatting behaviour by supplying his/her
// own formatting string, following the well-known <tt>printf</tt>-like
// conversions.
//
// \attention The user is responsible for the correctness of the optional
// conversion string. No checks are done by the toString() methods.

// @{

inline std::string toString(bool val)
{
  return val ? "true" : "false";
}

inline std::string toString(char val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  else return formatString("%hhi", val);
}

inline std::string toString(unsigned char val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%hhu", val);
}

inline std::string toString(short val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%hi", val);
}

inline std::string toString(unsigned short val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%hu", val);
}

inline std::string toString(int val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%i", val);
}

inline std::string toString(unsigned int val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%u", val);
}

inline std::string toString(long val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%li", val);
}

inline std::string toString(unsigned long val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%lu", val);
}

#if HAVE_LONG_LONG
inline std::string toString(long long val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%lli", val);
}

inline std::string toString(unsigned long long val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  return formatString("%llu", val);
}
#endif


inline std::string toString(float val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  else return formatString("%g", val);
}

inline std::string toString(double val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  else return formatString("%g", val);
}

#if HAVE_LONG_DOUBLE
inline std::string toString(long double val, const char* fmt = 0)
{
  if (fmt) return formatString(fmt, val);
  else return formatString("%Lg", val);
}
#endif

// @}

// @name Convert string to numeric value
// Convert a string to the value of any of the fundamental arithmetic data 
// types. Most of the StringTo...() methods provide the user with
// a means to override the default formatting behaviour by supplying his/her
// own formatting string, following the well-known <tt>scanf</tt>-like
// conversions.
// It uses sscanf, so a converting e.g. '2a' to int succeeds and results in 2.
//
// \attention The user is responsible for the correctness of the optional
// conversion string. No checks are done by the StringTo...() methods.
//
// \attention These functions will be deprecated in a next release.
// @{

bool   StringToBool  (const string& aString);      
int16  StringToInt16 (const string& aString,const char* fmt=0);
uint16 StringToUint16(const string& aString,const char* fmt=0);
int32  StringToInt32 (const string& aString,const char* fmt=0);
uint32 StringToUint32(const string& aString,const char* fmt=0);
#if HAVE_LONG_LONG
int64  StringToInt64 (const string& aString,const char* fmt=0);
uint64 StringToUint64(const string& aString,const char* fmt=0);
#endif
float  StringToFloat (const string& aString,const char* fmt=0);
double StringToDouble(const string& aString,const char* fmt=0);

// @}

// @name Strict conversion of string to numeric value
// Convert a string to the value of any of the fundamental arithmetic data 
// types. It checks if the entire string is used for the conversion.
// An integer value can also be given in hexadecimal format (e.g. 0x123).
// Leading and trailing whitespace is allowed.
// It checks if an integer value does not exceed the data type range.
// @{
long          strToLong   (const string& aString);
int           strToInt    (const string& aString);
int32         strToInt32  (const string& aString);
int16         strToInt16  (const string& aString);
unsigned long strToUlong  (const string& aString);
uint          strToUint   (const string& aString);
uint32        strToUint32 (const string& aString);
uint16        strToUint16 (const string& aString);
#if HAVE_LONG_LONG
int64         strToInt64  (const string& aString);
uint64        strToUint64 (const string& aString);
#endif
float         strToFloat  (const string& aString);
double        strToDouble (const string& aString);
inline bool   strToBool   (const string& aString)
  { return StringToBool (aString); }
// @}

// @name Manipulate strings containing a array specification
// Array specifications are often entered by the user with ranges like 3..32,55..58
// For converting such a string to a real vector the spec must be expanded so that
// it contains all elements instead of the ranges. 
// Likewise, when you present a array to the user you often want to show a spec with
// the ranges instead of all individual elements.
// See the ParameterSet document for a detailed description of the syntax.

// @{
// Given a string 'xx, xx, xx' this utility compacts the string
// by replacing series with range.
// Eg. [ lii001, lii002, lii003, lii005 ] --> [ lii001..lii003, lii005 ]
string compactedArrayString(const string&	orgStr);

// Given a string 'xx..xx, xx' this utility expands the string
// by replacing ranges with the fill series.
// Eg. [ lii001..003xx, lii005 ] --> [ lii001xx, lii002xx, lii003xx, lii005 ]
string expandRangeString(const string&);

// Given a string like '3*str' this utility expands the string
// by replacing the string 3 times.
// Eg. [3*0] --> [0,0,0]
string expandMultString(const string&);

// Apply expandMultString and expandRangeString (in that order) for an array
// string which must be enclosed in square brackets.
string expandArrayString(const string&);
// @} 

} // namespace LOFAR

#endif
