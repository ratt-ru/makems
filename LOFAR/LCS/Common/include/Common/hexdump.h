//#  hexdump.h: create hexdump of given datablock
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: hexdump.h 7330 2005-12-27 10:02:33Z loose $

#ifndef LOFAR_COMMON_HEXDUMP_H
#define LOFAR_COMMON_HEXDUMP_H

// \file
// Create hexdump of given datablock

//# Never #include <config.h> or #include <lofar_config.h> in a header file!
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>


namespace LOFAR
{

  void hexdump (						 	 const void* buf, int32	nrBytes);
  void hexdump (FILE*				aFile, 	 const void* buf, int32	nrBytes);
  void hexdump (char*				aChrPtr, const void* buf, int32	nrBytes);
  void hexdump (string&			aString, const void* buf, int32	nrBytes);
  
} // namespace LOFAR

#endif
