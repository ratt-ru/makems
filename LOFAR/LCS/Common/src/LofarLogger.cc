//#  LofarLogger.cc: Interface to the log4cplus logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.cc 11757 2008-09-17 09:49:47Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <cstdio>					// snprintf
#include <unistd.h>					// readlink
#include <libgen.h>				// basename
#include <cstring>
#include <Common/LofarLogger.h>

#if defined(HAVE_LOG4CXX)
# include <log4cxx/ndc.h>
#endif


namespace LOFAR {

//#------------------------- Internal implementation -----------------------------
//
// lofarLoggerInitNode()
//
// Creates a NDC with the text "application@node" and pushes it
// on the NDC stack
void lofarLoggerInitNode(void) {
	const int	MAXLEN = 128;
	int		applNameLen = 0;
	char	hostName [MAXLEN];
	char	applName [MAXLEN];
	char	loggerId [MAXLEN];

	// try to resolve the hostname
	if (gethostname (hostName, MAXLEN-1) < 0) {
		hostName[0]='\0';
	}

	// try to resolve the applicationname
	applNameLen = readlink("/proc/self/exe", applName, MAXLEN-1);
	if (applNameLen >= 0)
		applName[applNameLen] = '\0';
	else
		strcpy (applName, "");
//	}	

	// construct loggerId and register it.
	snprintf(loggerId, MAXLEN-1, "%s@%s", basename(applName), hostName);
#if defined(HAVE_LOG4CPLUS)
	log4cplus::getNDC().push(loggerId);
#elif defined(HAVE_LOG4CXX)
        log4cxx::NDC::push(loggerId);
#endif
}

#if defined(HAVE_LOG4CPLUS)
using namespace log4cplus;

//# ------------------------ implement the five trace levels ------------------------
const LogLevel	TRACE1_LOG_LEVEL	= 1;
const LogLevel	TRACE2_LOG_LEVEL	= 2;
const LogLevel	TRACE3_LOG_LEVEL	= 3;
const LogLevel	TRACE4_LOG_LEVEL	= 4;
const LogLevel	TRACE5_LOG_LEVEL	= 5;
const LogLevel	TRACE6_LOG_LEVEL	= 6;
const LogLevel	TRACE7_LOG_LEVEL	= 7;
const LogLevel	TRACE8_LOG_LEVEL	= 8;

#define _TRACE1_STRING	LOG4CPLUS_TEXT("TRACE1")
#define _TRACE2_STRING	LOG4CPLUS_TEXT("TRACE2")
#define _TRACE3_STRING	LOG4CPLUS_TEXT("TRACE3")
#define _TRACE4_STRING	LOG4CPLUS_TEXT("TRACE4")
#define _TRACE5_STRING	LOG4CPLUS_TEXT("TRACE5")
#define _TRACE6_STRING	LOG4CPLUS_TEXT("TRACE6")
#define _TRACE7_STRING	LOG4CPLUS_TEXT("TRACE7")
#define _TRACE8_STRING	LOG4CPLUS_TEXT("TRACE8")

tstring traceLevel2String(LogLevel ll) {
	switch (ll) {
	case TRACE1_LOG_LEVEL:	return _TRACE1_STRING;
	case TRACE2_LOG_LEVEL:	return _TRACE2_STRING;
	case TRACE3_LOG_LEVEL:	return _TRACE3_STRING;
	case TRACE4_LOG_LEVEL:	return _TRACE4_STRING;
	case TRACE5_LOG_LEVEL:	return _TRACE5_STRING;
	case TRACE6_LOG_LEVEL:	return _TRACE6_STRING;
	case TRACE7_LOG_LEVEL:	return _TRACE7_STRING;
	case TRACE8_LOG_LEVEL:	return _TRACE8_STRING;
	}

	return tstring();		// not found
}

LogLevel string2TraceLevel (const tstring& lname) {
	if (lname == _TRACE1_STRING)	return TRACE1_LOG_LEVEL;
	if (lname == _TRACE2_STRING)	return TRACE2_LOG_LEVEL;
	if (lname == _TRACE3_STRING)	return TRACE3_LOG_LEVEL;
	if (lname == _TRACE4_STRING)	return TRACE4_LOG_LEVEL;
	if (lname == _TRACE5_STRING)	return TRACE5_LOG_LEVEL;
	if (lname == _TRACE6_STRING)	return TRACE6_LOG_LEVEL;
	if (lname == _TRACE7_STRING)	return TRACE7_LOG_LEVEL;
	if (lname == _TRACE8_STRING)	return TRACE8_LOG_LEVEL;

	return NOT_SET_LOG_LEVEL;			// not found
}

LOFAR::LoggerReference theirTraceLoggerRef("TRC");   // create the tracelogger

#elif defined(HAVE_LOG4CXX)

log4cxx::LoggerPtr theirTraceLoggerRef(log4cxx::Logger::getLogger("TRC"));

#endif


// initTracemodule
//
// Function that is used when the TRACE levels are NOT compiled out. It registers
// the TRACEn management routines at the Log4Cplus LogLevelManager and sets up the
// global trace-logger named "TRACE", with the additivity set to false.
// Attached to the trace-logger is one Appender that logs to stderr.
//
void initTraceModule (void) {
#ifdef HAVE_LOG4CPLUS
	//# register our own loglevels
	getLogLevelManager().pushToStringMethod(traceLevel2String);
	getLogLevelManager().pushFromStringMethod(string2TraceLevel);
	
	//# Setup a property object to initialise the TRACE Logger
	helpers::Properties		traceProp;		
	traceProp.setProperty("log4cplus.logger.TRC", "DEBUG, STDERR");
	traceProp.setProperty("log4cplus.additivity.TRC", "false");
	traceProp.setProperty("log4cplus.appender.STDERR", "log4cplus::ConsoleAppender");
	traceProp.setProperty("log4cplus.appender.STDERR.logToStdErr", "true");
	traceProp.setProperty("log4cplus.appender.STDERR.ImmediateFlush", "true");
	traceProp.setProperty("log4cplus.appender.STDERR.layout","log4cplus::PatternLayout");
	traceProp.setProperty("log4cplus.appender.STDERR.layout.ConversionPattern",
							"%D{%y%m%d %H%M%S,%q} [%P] %-6p %c{3} [%F:%L] - %m%n");

	PropertyConfigurator(traceProp).configure();
	Logger::getInstance("TRC").forcedLog(0, "TRACE module activated");

#endif // HAVE_LOG4CPLUS
}

}	// namespace LOFAR
