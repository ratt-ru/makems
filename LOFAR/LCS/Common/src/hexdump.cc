//# Hexdump.cc: Produces hexdump of data block
//#
//# Copyright (C) 2002-2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: hexdump.cc 4847 2005-04-13 14:33:43Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/hexdump.h>
#include <Common/StringUtil.h>
#include <stdio.h>

using namespace LOFAR;
using namespace std;

namespace LOFAR
{

// Define hexdump modes
enum { HdmString, HdmFile, HdmCppString };

//
// internalHexdump(buf, nrBytes, mode)
//
void internalHexdump (void*			device, 
					  const void*	buf, 
					  int32 		nrBytes, 
					  int32			mode)
{
#define	BTS_P_LINE		16
#define HALF_LINE		(BTS_P_LINE/2)

	char  hex[(3*BTS_P_LINE)+4], chars[BTS_P_LINE+1];

	const unsigned char *ch = (const unsigned char*)buf;

	for (int32 left = 0; left < nrBytes; left += BTS_P_LINE) {
		for (int32 i = 0; i < BTS_P_LINE; i++, ch++) {
			if (left + i < nrBytes) {
				// add bytes code to hex and char string
				sprintf(hex+(i*3)+(i/HALF_LINE),"%02x  ",(int)*ch);
				chars[i] = (*ch>=32 && *ch<=127) ? *ch : '.';
			}
			else {
				sprintf(hex+(i*3)+(i/HALF_LINE), "    ");
				chars[i] = ' ';
			}
		}
		hex   [3*BTS_P_LINE+2] = '\0';
		chars [BTS_P_LINE] 	   = '\0';
		// show result
		switch (mode) {
		case HdmFile:
			fprintf (static_cast<FILE*>(device), "%04X: %s%s\n", left, hex, chars);
			break;
		case HdmString:
			sprintf (static_cast<char*>(device), "%04X: %s%s\n", left, hex, chars);
			device = static_cast<char*>(device) + (6+(3*BTS_P_LINE)+2+BTS_P_LINE+1);
			break;
		case HdmCppString:
			static_cast<string*>(device)->append(formatString("%04X: %s%s\n", left, hex, chars));
			break;
		}

		chars[0] = '\0';
	}
}

//
// hexdump(buf, nrBytes)		 to standard out
//
void hexdump (const void *buf, int32 	nrBytes)
{
	internalHexdump(stdout, buf, nrBytes, HdmFile);
}

//
// hexdump(file, buf, nrBytes)
//
void hexdump (FILE*	aFile, const void *buf, int32 	nrBytes)
{
	internalHexdump(aFile, buf, nrBytes, HdmFile);
}

//
// hexdump(string, buf, nrBytes)
//
void hexdump (char*	aString, const void *buf, int32 	nrBytes)
{
	internalHexdump(aString, buf, nrBytes, HdmString);
}

//
// hexdump(string&, buf, nrBytes)
//
void hexdump (string&	aString, const void *buf, int32 	nrBytes)
{
	internalHexdump(&aString, buf, nrBytes, HdmCppString);
}


} // namespace LOFAR
