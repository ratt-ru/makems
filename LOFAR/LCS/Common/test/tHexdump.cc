//#  testHexdump.cc: Program to test the hexdump functions
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tHexdump.cc 11938 2008-10-17 11:19:44Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/LofarLogger.h>	       // contains the API of the LofarLogger
#include <Common/hexdump.h>
#include <Common/StringUtil.h>
#include <cstring>

using namespace LOFAR;


int main (int, char *argv[]) {

	// Read in the log-environment configuration
	// We should always start with this.
	INIT_LOGGER("tHexdump");

	// Show operator were are on the air
	LOG_INFO (formatString("Program %s has started", argv[0]));

	LOG_INFO ("testing hexdump(stdout, ...)");
	char	stdoutTest [80];
	strcpy (stdoutTest, "this should appear on standard out");
	hexdump (stdoutTest, strlen(stdoutTest));

	LOG_INFO ("testing hexdump(FILE*, ...)");
	char	fileTest [80];
	strcpy (fileTest,"this should appear in the  file tHexdump_tmp.txt");
	FILE*	fd = fopen("./tHexdump_tmp.txt", "w");
	hexdump(fd, fileTest, strlen(fileTest));
	fclose(fd);

	LOG_INFO ("testing hexdump(char*, ...)");
	char	charPtrTest [80];
	char	hexdumpInfo [380];	// approx 4.8 * input string size
	strcpy(charPtrTest, "this should appear in the given string");
	hexdump(hexdumpInfo, charPtrTest, strlen(charPtrTest));
	cout << hexdumpInfo << endl;

	LOG_INFO ("testing hexdump(string, ...)");
	string		hexdumpString;
	char		stringTest [80];
	strcpy(stringTest, "this should appear in the C++ string");
	hexdump(hexdumpString, stringTest, strlen(stringTest));
	cout << hexdumpString << endl;

	LOG_INFO("Normal termination of program");
	return (0);
}

