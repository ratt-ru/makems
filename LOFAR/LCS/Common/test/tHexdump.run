#!/bin/sh

./tHexdump
STATUS=$?
cat tHexdump_tmp.txt
exit $STATUS

