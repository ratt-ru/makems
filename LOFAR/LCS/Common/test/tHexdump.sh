#!/bin/sh
./runctest.sh tHexdump 2>&1 > tHexdump.log
