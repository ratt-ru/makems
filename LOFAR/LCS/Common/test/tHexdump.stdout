0000: 74 68 69 73 20 73 68 6f  75 6c 64 20 61 70 70 65  this should appe
0010: 61 72 20 6f 6e 20 73 74  61 6e 64 61 72 64 20 6f  ar on standard o
0020: 75 74                                             ut              
0000: 74 68 69 73 20 73 68 6f  75 6c 64 20 61 70 70 65  this should appe
0010: 61 72 20 69 6e 20 74 68  65 20 67 69 76 65 6e 20  ar in the given 
0020: 73 74 72 69 6e 67                                 string          

0000: 74 68 69 73 20 73 68 6f  75 6c 64 20 61 70 70 65  this should appe
0010: 61 72 20 69 6e 20 74 68  65 20 43 2b 2b 20 73 74  ar in the C++ st
0020: 72 69 6e 67                                       ring            

0000: 74 68 69 73 20 73 68 6f  75 6c 64 20 61 70 70 65  this should appe
0010: 61 72 20 69 6e 20 74 68  65 20 20 66 69 6c 65 20  ar in the  file 
0020: 74 48 65 78 64 75 6d 70  5f 74 6d 70 2e 74 78 74  tHexdump_tmp.txt
